ModPeek/ParseZIP.cs

```diff
@@ -0,0 +1,277 @@
+﻿using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.IO.Compression;
+using Vintagestory.API.Common;
+
+namespace VintageStory.ModPeek;
+
+static partial class ModPeek
+{
+    // See tests for examples of the attributes we are trying to parse.
+    static bool TryGetZipInfo(byte[] bytes, out ModInfo? modInfo, Action<Error> errorCallback)
+    {
+        try
+        {
+            using var zip = new ZipArchive(new MemoryStream(bytes));
+            var modInfoEntry = zip.GetEntry("modinfo.json");
+            if (modInfoEntry == null) {
+                // Figure out if all entries start with the same path prefix. In that case the user zipped a folder instead of individual files.
+                string? commonPathPrefix = null;
+                foreach(var entry in zip.Entries) {
+                    var firstSeparator = entry.FullName.IndexOf('/');
+                    if (firstSeparator == -1) { // Indicates an individual file in the root, we are already good.
+                        commonPathPrefix = null;
+                        break;
+                    }
+
+                    if (commonPathPrefix == null) {
+                        commonPathPrefix = entry.FullName.Substring(0, firstSeparator + 1);
+                    }
+                    else if (!entry.FullName.StartsWith(commonPathPrefix)) { // Found other file / folder, so its not just one root directory.
+                        commonPathPrefix = null;
+                        break;
+                    }
+                }
+                errorCallback(Errors.MissingFileInArchiveRoot("modinfo.json", commonPathPrefix != null));
+                modInfo = null;
+                return false;
+            }
+
+            using var reader = new StreamReader(modInfoEntry.Open());
+            var content = reader.ReadToEnd();
+            try {
+                var root = JToken.Parse(content);
+                return TryParseJsonCaseInsensitive(root, out modInfo, errorCallback);
+            }
+            catch(Exception e) {
+                errorCallback(Errors.MalformedJson(e));
+                modInfo = null;
+                return false;
+            }
+        }
+        catch (Exception e)
+        {
+            errorCallback(Errors.MalformedArchive(e));
+            modInfo = null;
+            return false;
+        }
+    }
+
+    static bool TryParseJsonCaseInsensitive(JToken root, out ModInfo? modInfo, Action<Error> errorCallback)
+    {
+        if(root.Type != JTokenType.Object) {
+            errorCallback(Errors.UnexpectedJsonRootType(root, JTokenType.Object));
+            Console.Error.WriteLine($"The root json node must be an object, but was a {root.Type}.");
+            modInfo = null;
+            return false;
+        }
+
+        var error = false;
+        modInfo = new ModInfo();
+
+        foreach(var prop in ((JObject)root).Properties()) {
```

Comment:
can we avoid code duplication? (nearly the same code in ParseCS.cs)

Reply:
Well, they are similar, but one parses a json and the other parses an assembly attribute ast.
These are different structures, process completely different types and produce different errors, so I don't they ca reasonably be de-duplicated.